Genotype vogue

mayabrandi opened the pull request:
This PR adds cli to load data from genotype to vogue, using genotype and vogue CLIs




**How to test**:

On hasta
cd /home/proj/stage/servers/resources/hasta.scilifelab.se

1. bash update-servers-stage.sh genotype_vogue

2. bash update-genotype-stage.sh trending

3. bash update-cg-stage.sh genotype-vogue

4. bash update-vogue-stage.sh develop





4.  run: 
usestage
export MONGO_URI=mongodb://localhost:27030
export MONGO_DBNAME=vogue-stage
export FLASK_DEBUG=1
export VOGUE_SECRET_KEY=hej
cg -c /home/proj/stage/servers/config/hasta.scilifelab.se/cg-stage.yaml upload vogue genotype -d 300

**Expected outcome**:
Something like this:
----------------- UPLOAD ----------------------
----------------- TRENDING -----------------------
----------------- GENOTYPE -----------------------
vogue output: INFO:vogue.adapter.plugin:Use database vogue-stage.
vogue output: INFO:vogue.load.genotype:Trying to build json
vogue output: INFO:vogue.adapter.plugin:No updates for sample ADM1464A1.
...

**Review:**
- [x] code approved by @patrikgrenfeldt 
- [x] tests executed by @mayabrandi  9/12 - screen shot attached
- [x] "Merge and deploy" approved by @patrikgrenfeldt  and @emiliaol 
Thanks for filling in who performed the code review and the test!

This is a minor bump version since we are not making changes to existing functionality.

ingkebil commented:
Sooo.. Why do we need to create a cg.yaml file and not use the staging one?

ingkebil commented on cg/cli/upload.py:
```diff
@@ -452,3 +460,17 @@ def validate(context, family_id):
+@click.option('-d', '--days', 
+            help='load X days old sampels from genotype to vogue')
+@click.pass_context
+def trending_genotype(context, days: int):
```
So, I presume most of the vogue related transfers will be after an bioit analysis will be done, right? This function would upload as soon as as the analysis is finished and needs a `family_id` as parameter, not a "so many days ago" parameter. (Like the other functions ;))
We can then call it together with all the other upload functions from the `cg upload auto`.

We also have to come up with a better name than this. It has too many downsides: it uses an underscore, it uses two nouns. Most of the other upload functions answer the question: where are you going to upload (scout, genotype, ...) or what are you go upload (observations)? e.g. genotype-trends would work a lot better. If vogue is going to have more actions, why not add `cg upload trends` or `cg upload trends genotype` and even a 'cg upload trends all` (if that is even possible).

mayabrandi replied:
Like it is built, it can run every day and fetch what ever was updated in the database witin the last day eg. Or every second day, or whatever. Is that a problem? Otherwise Ill have to redo the genotype PR as well. Is it worth the amount of time? Would have been nice to get this kind of restrictions before starting to develop for the cg cli.

"cg uploads trends genotype"  I like! 👍

ingkebil replied:
Like discussed, let's drop the rewriting unless you really feel like it ;)

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +47,33 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
+            # If CalledProcessError is raised, log and raise error
+            log_msg = f"Could not run command: {' '.join(trending_call)}"
+            LOG.critical(log_msg)
+            raise
```
Please raise the original exception as well.

There is a good explanation on why to do so in "Clean Code for Python", page 60 and onward. Raising original exception is on page 68.

mayabrandi replied:
👍

mayabrandi replied:
I dont get what is the original exception here.. @ingkebil

ingkebil replied:
It's the `CalledProcessError`

ingkebil commented on .vscode/settings.json:
```diff
@@ -0,0 +1,6 @@
+{
```
please refrain from adding with `git add -A` and remove this file

ingkebil replied:
Same with the .DS_Store

mayabrandi replied:
👍

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +47,33 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
+    def get_trending(self, sample_id : str = '' , days: str = '' ):
+        """Get trending object with one or many samples."""
+        trending_obj = None
+        trending_call = copy.deepcopy(self.base_call)
```
There is no need to deepcopy a flat list

```suggestion
        trending_call = self.base_call[:]
```

ingkebil commented on cg/apps/vogue.py:
```diff
@@ -0,0 +1,62 @@
+
+    def load_genotype(self, genotype_dict: dict ):
+        """Add observations from a VCF."""
+        load_call = copy.deepcopy(self.base_call)
```
```suggestion
        load_call = self.base_call[:]
```

ingkebil commented on cg/apps/vogue.py:
```diff
@@ -0,0 +1,62 @@
+        for line in execute_command(load_call):
+            log_msg = f"vogue output: {line}"
+            LOG.info(log_msg)
+            line_content = line.split('INFO')[-1].strip()
```
... why do we need this line?

mayabrandi replied:
we don't :)

ingkebil replied:
Let's get rid of it :)

ingkebil commented on cg/meta/upload/vogue.py:
```diff
@@ -0,0 +1,22 @@
+from cg.apps.gt import GenotypeAPI
+from cg.apps.vogue import VogueAPI
+import json
+from django.core.serializers.json import DjangoJSONEncoder
```
eh .. could we use something less dramatic than django?

mayabrandi replied:
Ops. I wasn't using that anymore. Removing the import.

ingkebil commented:
Additional remarks:
- please add some tests :)
- the build is failing, mostly on a shitload of linting errors. Can you have a look at that?

mayabrandi commented:
Didn't get time to add tests today. Nor reading up about assertionErrors. Believe I fixed most of the rest. See you monday @ingkebil

mayabrandi commented:
Tests are added. There is one lint error that I don't get: 
'check_output' has no 'assert_called_with' member
You know what it means? @ingkebil

ingkebil commented:
Nope, and I don't really understand why it's happening. As this is part of the tests and the tests pass, than I'd say just ignore it. You could add a pylint exception so the linter stops complaining about it.

In step 4. of how to test you say: create cg.yaml. Can you add a proper cg.yaml and cg-stage.yaml to the servers repo? We cannot push this to production if we don't have the proper cg.yaml available.

Feel free to test and document the test yourself btw! :)

mayabrandi commented:
I added a cg-stage.yaml and a update-genotype-stage.sh to servers. Since there is no existing genotype environment, I added one for the purpose of testig this PR. But i need help reviewing the update script. can you have a look @ingkebil? https://github.com/Clinical-Genomics/servers/tree/genotype_vogue

ingkebil commented:
Why don't you make a PR?

ingkebil commented on cg/apps/gt.py:
```diff
@@ -1,6 +1,11 @@
+from subprocess import CalledProcessError
+import subprocess
+
+from cg.exc import CaseNotFoundError
```
It's a trivial remark, the way imports are ordered is the following:
- build ins
- 3rd party
- internal

Can you move the cg import down to the internal imports.

Source: [PEP8](https://www.python.org/dev/peps/pep-0008/#imports)

mayabrandi replied:
👍

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +46,31 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
                 analysis_obj.sample.sex = samples_sex[analysis_obj.sample_id]['pedigree']
                 self.commit()
+
+    def get_trending(self, sample_id: str = '', days: str = '') -> dict:
```
seems weird to have days as a str and not an int. Any reason for that?

mayabrandi replied:
Nope.

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +46,31 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
+        try:
+            LOG.info('Running Genotype API to get data.')
+            output = subprocess.check_output(
+                ' '.join(trending_call),
```
There is no need to join this. Subprocess does that for you.
```suggestion

```

mayabrandi replied:
I dont get it to work when removing the join though. 

  File "/Users/mayabrandi/opt/cg/cg/meta/upload/vogue.py", line 18, in load_genotype
    samples = json.loads(samples)
  File "/Users/mayabrandi/anaconda/envs/cg/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/Users/mayabrandi/anaconda/envs/cg/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/Users/mayabrandi/anaconda/envs/cg/lib/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
(cg) 130-229-1-124-dhcp:cg mayabrandi$

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +46,31 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
+                shell=True
+            )
+        except CalledProcessError as error:
+            # If CalledProcessError is raised, log and raise error
```
No need for superfluous comment. It looks pretty obvious what is happening ;)
```suggestion

```

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +46,31 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
+            )
+        except CalledProcessError as error:
+            # If CalledProcessError is raised, log and raise error
+            log_msg = f"Could not run command: {' '.join(trending_call)}"
```
Don't build a log message before, instead pass the log message as a an `%` string

mayabrandi replied:
👍

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +46,31 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
+        output = output.decode('utf-8')
+        # If sample not in genotype db, stdout of genotype command will be empty.
+        if not output:
+            raise CaseNotFoundError(f"samples not found in genotype db")
```
Nice! :)

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +46,31 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
+            raise error
+
+        output = output.decode('utf-8')
+        # If sample not in genotype db, stdout of genotype command will be empty.
```
Here the comment is actually helpful as it explains why the error is raised

ingkebil commented on cg/apps/vogue.py:
```diff
@@ -0,0 +1,58 @@
+
+import json
+import logging
+
```
subprocess is a built in, no need for a space.
```suggestion
```

ingkebil commented on cg/apps/vogue.py:
```diff
@@ -0,0 +1,58 @@
+
+        # Execute command and print its stdout+stderr as it executes
+        for line in execute_command(load_call):
+            log_msg = f"vogue output: {line}"
```
Don't use f-string for log messages. Please use `%` strings instead

ingkebil replied:
This.

mayabrandi replied:
Okej!

ingkebil commented on cg/apps/vogue.py:
```diff
@@ -0,0 +1,58 @@
+        yield line.decode('utf-8').strip()
+
+    # Check if process exited with returncode != 0
+    if process.poll():
```
Can you be more explicit about this condition?
```suggestion
    if process.poll() != 0:
```

ingkebil replied:
This.

mayabrandi replied:
Ah. Sorry. Missed that one to. Is this better?

ingkebil commented on cg/cli/trending/base.py:
```diff
@@ -0,0 +1,14 @@
+"""Base command for trending"""
+
+import click
```
```suggestion
import click

```

ingkebil commented on cg/apps/vogue.py:
```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
```
If you save as uft-8, the file is utf-8 ;)

```suggestion

```

ingkebil replied:
Unless it has another purpose?

mayabrandi replied:
👍

ingkebil commented on cg/cli/upload.py:
```diff
@@ -78,6 +88,8 @@ def upload(context, family_id):
             click.echo(click.style(f"{family_id}: analysis uploaded!", fg='green'))
 
 
+upload.add_command(trending_command)
```
Please add this to the end of the file instead. No one is expecting an add_command between function definitions ;)

![image](https://user-images.githubusercontent.com/239608/65821799-c99cfc80-e23a-11e9-893a-ad9655d0d16d.png)

mayabrandi replied:
:) okej

ingkebil commented on cg/meta/upload/vogue.py:
```diff
@@ -0,0 +1,21 @@
+"""API to run Vogue"""
+# -*- coding: utf-8 -*-
+import json
```
```suggestion
import json

```

ingkebil commented on tests/apps/gt/conftest.py:
```diff
@@ -0,0 +1,34 @@
+
+    _config = CONFIG
+
+    return _config
```
Nice!

ingkebil commented on tests/apps/gt/test_gt_api.py:
```diff
@@ -0,0 +1,70 @@
+import pytest
+
+from cg.apps.gt import GenotypeAPI
+
```
```suggestion
```

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +46,31 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
+        if sample_id:
+            trending_call.extend(['prepare-trending', '-s', sample_id])
+        elif days:
+            trending_call.extend(['prepare-trending', '-d', days])
```
Is it ok to have no sample and no days?

mayabrandi replied:
Nope. How should I handle it differently?

ingkebil commented on cg/apps/gt.py:
```diff
@@ -37,3 +46,31 @@ def upload(self, bcf_path: str, samples_sex: dict, force: bool=False):
                 analysis_obj.sample.sex = samples_sex[analysis_obj.sample_id]['pedigree']
                 self.commit()
+
+    def get_trending(self, sample_id: str = '', days: str = '') -> dict:
```
Looks like the return value is a byte string, not a dict 

https://docs.python.org/2/library/subprocess.html#subprocess.check_output

mayabrandi replied:
correct

ingkebil commented on cg/apps/vogue.py:
```diff
@@ -0,0 +1,58 @@
+        self.vogue_binary = config['vogue']['binary_path']
+        self.base_call = [self.vogue_binary]
+
+    def load_genotype(self, genotype_dict: dict):
```
Are you sure this is a dict?

mayabrandi replied:
actually yes

ingkebil commented on cg/cli/trending/__init__.py:
```diff
@@ -0,0 +1,2 @@
+"""Init for the trending CLI"""
```
nice :)

ingkebil commented on cg/cli/trending/genotype.py:
```diff
@@ -0,0 +1,19 @@
+@click.option('-d', '--days',
+              help='load X days old sampels from genotype to vogue')
+@click.pass_context
+def genotype(context, days: int):
```
hehe, days is an int here ;)

mayabrandi replied:
:)

mayabrandi commented:
@ingkebil I don't get to set the envitonments up properly. Running bash update-servers-stage.sh genotype_vogue does not result in the update script for vogue to show up. Need help

mayabrandi commented:
When running the test on hasta according to specifications above, i get:

![Skärmavbild 2019-10-10 kl  14 41 19](https://user-images.githubusercontent.com/1306333/66569551-16060780-eb6c-11e9-8c5c-a241972e093e.png)

The error comes from datetime. [here](https://github.com/Clinical-Genomics/vogue/blob/50a2cb3ed3c4efcf33c8477427bca7c31cc1e580/vogue/load/genotype.py#L14) 

Running the same thing locally with the same installations i don't get the datetime error. Im running with the same python version as on hasta. Do you have a clue why this could be? @ingkebil 

![Skärmavbild 2019-10-10 kl  14 40 27](https://user-images.githubusercontent.com/1306333/66569501-ff5fb080-eb6b-11e9-8f2d-ea568f528ffe.png)

ingkebil commented:
Asto the error: my best guess is the data provided locally only holds a `%Y-%m-%d` formatted date, while in production we have it formatted as `%Y-%m-%dT%H%M%S`

mayabrandi commented:
Okej. Thanks.

> 10 okt. 2019 kl. 14:52 skrev Kenny Billiau <notifications@github.com>:
> 
> I wouldn't use datetime to convert dates and instead rely on dateutils: https://dateutil.readthedocs.io/en/stable/ <https://dateutil.readthedocs.io/en/stable/>
> It's a lot more stable as you do not need to care on how the date is formatted, dateutils will figure it out for you and return a datetime object.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Clinical-Genomics/cg/pull/415?email_source=notifications&email_token=AAJ65XLIZZXWXQYIH7ZUSWDQN4QRXA5CNFSM4IS3QEDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEA4E6AY#issuecomment-540561155>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJ65XLUKFSOL7APZKO4RYLQN4QRXANCNFSM4IS3QEDA>.
>

mayabrandi commented:
Its the same data though. Since im working against the same databases from locally as from hasta. So i don’t get how it can be different.


> 10 okt. 2019 kl. 14:54 skrev Kenny Billiau <notifications@github.com>:
> 
> Asto the error: my best guess is the data provided locally only holds a %Y-%m-%d formatted date, while in production we have it formatted as %Y-%m-%dT%H%M%S
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/Clinical-Genomics/cg/pull/415?email_source=notifications&email_token=AAJ65XMT7745KPLHN5RIYNDQN4QXNA5CNFSM4IS3QEDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEA4FFYY#issuecomment-540562147>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJ65XJLZ43PTURHOPCM25LQN4QXNANCNFSM4IS3QEDA>.
>

mayabrandi commented:
Code has now been tested on hasta with successful result. 
![Skärmavbild 2019-10-11 kl  09 59 33](https://user-images.githubusercontent.com/1306333/66634775-148d1b80-ec0e-11e9-9fea-5e54cb990bae.png)

ingkebil commented:
Almost there. Can you check the code for f-strings used in log messages and use %-strings instead? I found one where I made a remark.

If you don't find my comments back, the easiest is to look at the 'Files Changed' tab and scroll through. It's easy to spot everything that is not a resolved conversation.

mayabrandi commented:
> Almost there. Can you check the code for f-strings used in log messages and use %-strings instead? I found one where I made a remark.
> 
I searched for all f” and changed to %-strings instead.
> If you don't find my comments back, the easiest is to look at the 'Files Changed' tab and scroll through. It's easy to spot everything that is not a resolved conversation.
> 

Thanks for pointing out

ingkebil commented on tests/apps/gt/test_gt_api.py:
```diff
@@ -35,7 +35,7 @@ def test_get_trending(genotypeapi, mocker):
 
     # THEN assert subprocess is running the GenotypeAPI with correct command
-    call = f"gtdb --database database prepare-trending -d {days}"
+    call = "gtdb --database database prepare-trending -d %s" % days
```
It's really fine to leave the f-string here ;) You do not want to have the f-string in a log message as those only get interpolated if we need them.

The linter would tell you the same thing

mayabrandi replied:
okej!

ingkebil commented:
> Almost there. Can you check the code for f-strings used in log messages and use %-strings instead? I found one where I made a remark.
> I searched for all f” and changed to %-strings instead.

That's not what I mean. Don't use f-strings for log messages. Please harvest the power of f-strings everywhere else :)

patrikgrenfeldt commented on cg/apps/vogue.py:
```diff
@@ -0,0 +1,60 @@
+        self.vogue_binary = config['vogue']['binary_path']
+        self.base_call = [self.vogue_binary]
+
+    def load_genotype(self, genotype_dict: dict):
```
I suggest renaming this method to add_observations_from(self, vcf: dict) according to the docstring, this to avoid hard and soft couplings between our different applications, what do you say?

patrikgrenfeldt commented on cg/cli/upload.py:
```diff
@@ -507,3 +516,5 @@ def processed_solved(context):
     mutacc_auto_api = mutacc_auto.MutaccAutoAPI(context.obj)
     mutacc_auto_api.import_reads()
+
+upload.add_command(trending_command)
```
I suggest we move this to cli\base.py

mayabrandi replied:
👍

mayabrandi commented on cg/apps/gt.py:
```diff
@@ -21,6 +25,10 @@ def __init__(self, config: dict):
         alchy_config = dict(SQLALCHEMY_DATABASE_URI=config['genotype']['database'])
         super(GenotypeAPI, self).__init__(config=alchy_config, Model=models.Model)
 
+        self.genotype_database = config['genotype']['database']
```
could we use the config rather the specifics in the config

ingkebil replied:
Yes, that's definitely the way to go. Loqusdb does this already so you'll find an example on how to do this for cg and for the config

patrikgrenfeldt replied:
I've checked now and there is a config for genotype, please consider using it like Adam did in Loqus

patrikgrenfeldt commented:
@mayabrandi , I like that you added tests for the cg part of this implementation, how do you think about genotype?

ingkebil commented on cg/apps/vogue.py:
```diff
@@ -0,0 +1,60 @@
+        # Execute command and print its stdout+stderr as it executes
+        for line in execute_command(load_call):
+            log_msg = "vogue output: %s" % line
+            LOG.info(log_msg)
```
also the format of logging is not `str % var`, rather it is `str, var`. Eventhoug it looks like a %-string, it is not. It's logging module's own way of formatting.

mayabrandi commented:
<img width="1279" alt="Skärmavbild 2019-12-09 kl  14 17 47" src="https://user-images.githubusercontent.com/1306333/70439091-84454880-1a8f-11ea-8bf7-33dae4e5c350.png">

mayabrandi commented:
Can this one be approved and merged now @patrikgrenfeldt?

ingkebil commented:
It says in the description that this is amajor version bump, which requires a validation. Has the validation been performed?

henrikstranneheim commented:
It looks like we are only adding features and not changing existing API:s. Therefore, I think this should be a minor version bump. @patrikgrenfeldt what do you think?

patrikgrenfeldt commented:
I think this could be a minor, we are not making changes to existing functionality as far as I understand. I think that we should wait until Maya is present to deploy since we update many applications as configurations. It might be helpful to have you nearby in case of some small complications, if we have big ones we can always rollback, or what do you think @mayabrandi?

mayabrandi commented:
Perfect. Ill be there tomorrow and friday.

> 11 dec. 2019 kl. 12:41 skrev Patrik Grenfeldt <notifications@github.com>:
> 
> I think this could be a minor, we are not making changes to existing functionality as far as I understand. I think that we should wait until Maya is present to deploy since we update many applications as configurations. It might be helpful to have you nearby in case of some small complications, if we have big ones we can always rollback, or what do you think @mayabrandi <https://github.com/mayabrandi>?
> 
>

patrikgrenfeldt commented:
Great, let's deploy when you get here then!

patrikgrenfeldt commented:
deployed
